src/object_store/src/object/sim/service.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use bytes::Bytes;
+use futures::Stream;
+use itertools::Itertools;
+use spin::Mutex;
+
+use crate::object::error::ObjectResult as Result;
+use crate::object::sim::SimError;
+use crate::object::{ObjectError, ObjectMetadata};
+
+#[derive(Debug)]
+pub(crate) enum Request {
+    Upload { path: String, obj: Bytes },
+    Read { path: String },
+    Delete { path: String },
+    List { path: String },
+    Metadata { path: String },
+}
+
+#[derive(Debug)]
+pub(crate) enum Response {
+    Upload,
+    Read(Bytes),
+    Delete,
+    List(SimObjectIter),
+    Metadata(ObjectMetadata),
+}
+
+#[derive(Debug)]
+pub(crate) struct SimService {
+    storage: Mutex<BTreeMap<String, (ObjectMetadata, Bytes)>>,
+}
+
+fn get_obj_meta(path: &str, obj: &Bytes) -> Result<ObjectMetadata> {
+    Ok(ObjectMetadata {
+        key: path.to_string(),
+        last_modified: SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(ObjectError::internal)?
+            .as_secs_f64(),
+        total_size: obj.len(),
+    })
+}
+
+impl SimService {
+    pub fn new() -> Self {
+        SimService {
+            storage: Mutex::new(BTreeMap::new()),
+        }
+    }
+
+    pub async fn upload(&self, path: String, obj: Bytes) -> Result<Response> {
+        let metadata = get_obj_meta(&path, &obj)?;
+        self.storage.lock().insert(path.into(), (metadata, obj));
+        Ok(Response::Upload)
+    }
+
+    pub async fn read(&self, path: String) -> Result<Response> {
+        let storage = self.storage.lock();
+        let obj = storage
+            .get(&path)
+            .map(|(_, o)| o)
+            .ok_or_else(|| SimError::NotFound(format!("no object at path '{}'", path)))?;
+        Ok(Response::Read(obj.clone()))
+    }
+
+    pub async fn delete(&self, path: String) -> Result<Response> {
+        self.storage.lock().remove(&path);
+        Ok(Response::Delete)
+    }
+
+    pub async fn list(&self, prefix: String) -> Result<Response> {
+        let list_result = self
+            .storage
+            .lock()
+            .iter()
+            .filter_map(|(path, (metadata, _))| {
+                if path.starts_with(&prefix) {
+                    return Some(metadata.clone());
+                }
```

Comment:
We can use range scan `BTreeMap::range(prefix..prefix + 1)` to avoid the full scan. 
(where `prefix + 1` means to replace the last char `c` with `c+1`)